gen/constant.go

```diff
@@ -20,17 +20,222 @@
+				<end>
+			}`, struct {
+				Spec  *compile.ListSpec
+				Value compile.ConstantList
+			}{Spec: l, Value: v},
+			TemplateFunc("constantValue", ConstantValue))
+	case compile.ConstantMap:
+		m, ok := t.(*compile.MapSpec)
+		if !ok {
+			return "", fmt.Errorf(
+				"cannot cast map %v to %v", v, t.ThriftName())
+		}
+		return g.TextTemplate(
+			`
+			<$keyType := .Spec.KeySpec>
+			<$valueType := .Spec.ValueSpec>
+			<typeReference .Spec>{
+				<range .Value>
+					<if isHashable $keyType>
+						<constantValue .Key $keyType>:
+							<constantValue .Value $valueType>,
+					<else>
+						{
+							Key: <constantValue .Key $keyType>,
+							Value: <constantValue .Value $valueType>,
+						},
+					<end>
+				<end>
+			}`, struct {
+				Spec  *compile.MapSpec
+				Value compile.ConstantMap
+			}{Spec: m, Value: v},
+			TemplateFunc("constantValue", ConstantValue))
+	case compile.ConstantSet:
+		s, ok := t.(*compile.SetSpec)
+		if !ok {
+			return "", fmt.Errorf(
+				"cannot cast set %v to %v", v, t.ThriftName())
+		}
+		return g.TextTemplate(
+			`
+			<$valueType := .Spec.ValueSpec>
+			<typeReference .Spec>{
+				<range .Value>
+					<if isHashable $valueType>
+						<constantValue . $valueType>: struct{}{},
+					<else>
+						<constantValue . $valueType>,
+					<end>
+				<end>
+			}`, struct {
+				Spec  *compile.SetSpec
+				Value compile.ConstantSet
+			}{Spec: s, Value: v},
+			TemplateFunc("constantValue", ConstantValue))
+	case compile.ConstantString:
+		return strconv.Quote(string(v)), nil
+	case *compile.ConstantStruct:
+		s, ok := t.(*compile.StructSpec)
+		if !ok {
+			return "", fmt.Errorf(
+				"cannot cast struct %v to %v", v, t.ThriftName())
+		}
+		return g.TextTemplate(
+			`
+			<$fields := .Spec.Fields>
+			&<typeName .Spec>{
+				<range $name, $value := .Value.Fields>
+					<$field := $fields.FindByName $name>
+					<if and (not $field.Required) (isPrimitiveType $field.Type)>
+						<goCase $field.Name>: <primitiveValueRef $value $field.Type>,
+					<else>
+						<goCase $field.Name>: <constantValue $value $field.Type>,
+					<end>
+				<end>
+			}`, struct {
+				Spec  *compile.StructSpec
+				Value *compile.ConstantStruct
+			}{Spec: s, Value: v},
+			TemplateFunc("constantValue", ConstantValue),
+			TemplateFunc("primitiveValueRef", primitiveValueRef),
+		)
+	case compile.EnumItemReference:
+		return g.TextTemplate(`<typeName .Enum><goCase .Item.Name>`, v)
+	case compile.ConstReference:
+		return g.LookupConstantName(v.Target)
+	default:
+		panic(fmt.Sprintf("Unknown constant value %v (%T)", c, c))
+	}
+}
+
+// helper to generate pointers to primitives
+func primitiveValueRef(g Generator, c compile.ConstantValue, t compile.TypeSpec) (string, error) {
+	var toptr string
+	switch t {
+	case compile.BoolSpec:
+		toptr = "_boolptr"
+		err := g.EnsureDeclared(
+			`func _boolptr(v bool) *bool { return &v }`, nil)
+		if err != nil {
+			return "", err
+		}
+	case compile.I8Spec:
+		toptr = "_i8ptr"
```

Comment:
optional: since we parse the template and get the function declaration we find, should `EnsureDeclared` just return the name rather than duplicating it?